leopard.nim

```diff
@@ -0,0 +1,329 @@
+
+func isValid*(code: ReedSolomonCode): bool =
+  not ((code.codeword - code.data != code.parity) or
+       (code.parity > code.data) or (code.codeword < MinSymbols + 1) or
+       (code.data < MinSymbols) or (code.parity < MinSymbols) or
+       (code.codeword > MaxTotalSymbols))
+
+# alloc/freeAligned and helpers adapted from mratsim/weave:
+# https://github.com/mratsim/weave/blob/master/weave/memory/allocs.nim
+
+func isPowerOfTwo(n: int): bool {.inline.} =
+  (n and (n - 1)) == 0
+
+func roundNextMultipleOf(x: Natural, n: Natural): int {.inline.} =
+  (x + n - 1) and not (n - 1)
+
+when defined(windows):
+  proc aligned_alloc_windows(size, alignment: csize_t): pointer
+    {.header: "<malloc.h>", importc: "_aligned_malloc", sideeffect.}
+
+  proc aligned_free_windows(p: pointer)
+    {.header: "<malloc.h>", importc: "_aligned_free", sideeffect.}
+
+  proc freeAligned*(p: pointer) =
+    if not p.isNil:
+      aligned_free_windows(p)
+
+elif defined(osx):
+  proc posix_memalign(mem: var pointer, alignment, size: csize_t)
+    {.header: "<stdlib.h>", importc, sideeffect.}
+
+  proc aligned_alloc(alignment, size: csize_t): pointer {.inline.} =
+    posix_memalign(result, alignment, size)
+
+else:
+  proc aligned_alloc(alignment, size: csize_t): pointer
+    {.header: "<stdlib.h>", importc, sideeffect.}
+
+when not defined(windows):
+  proc c_free(p: pointer) {.header: "<stdlib.h>", importc: "free".}
+
+  proc freeAligned*(p: pointer) {.inline.} =
+    if not p.isNil:
+      c_free(p)
+
+proc allocAligned*(size: int, alignment: static Natural): pointer {.inline.} =
+  ## aligned_alloc requires allocating in multiple of the alignment.
+  static:
+    assert alignment.isPowerOfTwo()
+
+  let
+    requiredMem = size.roundNextMultipleOf(alignment)
+
+  when defined(windows):
+    aligned_alloc_windows(csize_t requiredMem, csize_t alignment)
+  else:
+    aligned_alloc(csize_t alignment, csize_t requiredMem)
+
+proc leoInit*() =
+  if wrapper.leoInit() != 0:
+    raise (ref LeopardDefect)(msg: "Leopard-RS failed to initialize")
+
+proc encode*(code: ReedSolomonCode, data: Data):
+    Result[ParityData, LeopardError] =
+  if not code.isValid:
+    return err LeopardError(code: LeopardBadCode, msg: LeopardBadCodeMsg)
+
+  var
+    data = data
+
+  let
+    symbolBytes = data[0].len
+
+  if data.len < code.data.int:
+    return err LeopardError(code: LeopardNotEnoughData,
+      msg: LeopardNotEnoughDataMsg)
+
+  elif data.len > code.data.int:
+    return err LeopardError(code: LeopardTooMuchData,
+      msg: $leoResultString(wrapper.LeopardTooMuchData))
+
+  if symbolBytes < MinBufferSize.int or symbolBytes mod MinBufferSize.int != 0:
+    return err LeopardError(code: LeopardInvalidSize,
+      msg: $leoResultString(wrapper.LeopardInvalidSize))
+
+  var
+    enData = newSeq[pointer](code.data)
+
+  for i in 0..<code.data:
+    if data[i].len != symbolBytes:
+      for i in 0..<code.data: freeAligned enData[i]
```

Comment:
This is wrong - calling any flavor of free would imply that you won't use the buffer anymore.

Reply:
It seems to be correct in the current implementation, though keep in mind what I wrote above re: working on a refactor.

In the current implementation, the `if` of L166 checks if there is a condition under which `proc encode` will return early. Whether returning early or at the very end, memory allocated with `allocAligned` needs to be freed because it's only used within the body of the proc, never outside of it. In this case `enData` is the `seq[pointer]` that holds the data copied from the `data` parameter, where each pointer in `enData` points to the start of memory allocated with `allocAligned`.

Reply:
But you havent called `allocAligned` by this point yet, that only happens in L171, calling `free` without calling `alloc` is at best undefined behaviour, but quite possibly you're freeing someone elses memory if the pointer happens to point to some valid region (i.e. has some garbage/isn't initialized to `nil`).

Reply:
Not quite. In [`L166-L174`](https://github.com/status-im/nim-leopard/blob/initial_impl/leopard.nim#L166-L174), the allocations are done in a `for` loop. The "early return" could happen at any step in the loop; any memory already allocated needs to be freed and null pointers would have been skipped because `freeAligned` skips them. In the latest revision, null pointers are skipped outright without depending on the "skip nulls" behavior of `freeAligned`.

Per discussion on Discord, the point is moot, since this PR will be closed.